### PR TITLE
feat(elements): Update exports and prep for public builds

### PR DESCRIPTION
**Pull request by tmilewski**
## Description

95% of this PR is simply moving components out of single `index.tsx` files and updating imports and the specific imported component names.

**_Adding comments to call-out specific points of interest._**

- Splits out components into their respective files from one large `index.tsx`.
- Enable export subpaths for:
  - `/common`
  - `/sign-in`
  - `/sign-up`
- Enable the ability to publish builds.
- Rename component exports to better handle the case where we don't want to use `import * as ___ from ___`
- Adds analysis capabilites via `npm run build:analyze`

---

Components can now be imported into applications using the following structures:

```tsx
import { GlobalError, Submit } from '@clerk/elements/common';
import { Continue, Factor, SignIn, SocialProvider, Start, Verification } from '@clerk/elements/sign-in';
```

```tsx
import * as Clerk from '@clerk/elements/common';
import * as SignIn from '@clerk/elements/sign-in';
```

Note: `Root` also maps to `SignIn` allowing for `SignIn.Root` rather than `SignIn.SignIn`

Screenshot 2024-01-29 at 4 40 10 PM

## Checklist

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerk/clerk-docs) has been updated

## Type of change

- [x] 🐛 Bug fix
- [x] 🌟 New feature
- [x] 🔨 Breaking change
- [x] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [ ] `@clerk/backend`
- [ ] `@clerk/chrome-extension`
- [ ] `@clerk/clerk-js`
- [ ] `@clerk/clerk-expo`
- [ ] `@clerk/fastify`
- [ ] `gatsby-plugin-clerk`
- [ ] `@clerk/localizations`
- [ ] `@clerk/nextjs`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/remix`
- [ ] `@clerk/clerk-sdk-node`
- [ ] `@clerk/shared`
- [ ] `@clerk/themes`
- [ ] `@clerk/types`
- [ ] `build/tooling/chore`

### CODE REVIEW

File: packages/elements/package.json [hunk 1]
```diff
@@ -49,6 +61,8 @@
     "app:e2e": "(cd examples/nextjs && npm run e2e)",
     "app:lint": "(cd examples/nextjs && npm run lint)",
     "build": "tsup",
+    "build:analyze": "tsup --metafile; open https://esbuild.github.io/analyze/",
```
Comment (tmilewski) [inline packages/elements/package.json:64]: Enables easy analysis of generated packages.

File: packages/elements/src/index.ts [hunk 2]
```diff
@@ -1,37 +1 @@
...
-  SignUpSocialProvider,
-  SignUpSocialProviderIcon,
-  SignUpStrategy,
-} from '~/react/sign-up';
-
-/** Hooks */
-export { useSignUpFlow, useSignUpFlowSelector } from '~/internals/machines/sign-up/sign-up.context';
-export { useSignInFlow, useSignInFlowSelector } from '~/internals/machines/sign-in/sign-in.context';
-export { useNextRouter };
+throw new Error('Please import specific subpaths instead. See documentation for more details: https://clerk.com/docs/'); // TODO: Add link to docs
```
Comment (tmilewski) [inline packages/elements/src/index.ts:1]: As-is, we don't need to make anything available at the top-level of the package.

File: packages/elements/tsup.config.ts [hunk 3]
```diff
@@ -17,6 +17,9 @@ export default defineConfig(overrideOptions => {
     dts: true,
     entry: {
       index: 'src/index.ts',
+      'react/common/index': 'src/react/common/index.ts',
+      'react/sign-in/index': 'src/react/sign-in/index.ts',
+      'react/sign-up/index': 'src/react/sign-up/index.ts',
```
Comment (tmilewski) [inline packages/elements/tsup.config.ts:22]: Create specific entrypoints

File: packages/elements/package.json [hunk 4]
```diff
@@ -26,6 +26,16 @@
   "author": "Clerk",
   "sideEffects": false,
   "exports": {
+    "./*": {
```
Comment (tmilewski) [inline packages/elements/package.json:29]: All React-specific code is contained within `dist/react` but available at the top-level by default, per previous conversations.

File: packages/elements/examples/nextjs/components/form.tsx [hunk 5]
```diff
@@ -43,6 +43,7 @@ export const CustomField = forwardRef<Input, { name: string; label: string }>(fu
         }
       : {
           className: 'bg-tertiary rounded-sm px-2 py-1 border border-foreground data-[invalid]:border-red-500',
+          ref: forwardedRef,
```
Comment (tmilewski) [inline packages/elements/examples/nextjs/components/form.tsx:46]: `forwardedRef` only applies to the non-`otp` input at this point.

File: packages/elements/package.json [hunk 6]
```diff
@@ -59,14 +73,12 @@
...
-    "@clerk/clerk-react": "5.0.0-beta-v5.17",
-    "@clerk/nextjs": "5.0.0-beta-v5.20",
-    "@clerk/shared": "2.0.0-beta-v5.11",
+    "@clerk/clerk-react": "^5.0.0-beta-v5.17",
+    "@clerk/shared": "^2.0.0-beta-v5.11",
     "@radix-ui/react-form": "^0.0.3",
     "@radix-ui/react-slot": "^1.0.2",
     "@statelyai/inspect": "^0.1.0",
     "@xstate/react": "^4.0.2",
-    "clsx": "^2.1.0",
```
Comment (tmilewski) [inline packages/elements/package.json:69]: Wasn't used.

**Comment by tmilewski** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @tmilewski - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/backend | 1.0.1-snapshot.v59b1e3b |
| @clerk/chrome-extension | 1.0.1-snapshot.v59b1e3b |
| @clerk/clerk-js | 5.0.1-snapshot.v59b1e3b |
| @clerk/elements | 0.0.2-snapshot.v59b1e3b |
| @clerk/clerk-expo | 1.0.1-snapshot.v59b1e3b |
| @clerk/fastify | 1.0.1-snapshot.v59b1e3b |
| gatsby-plugin-clerk | 5.0.1-snapshot.v59b1e3b |
| @clerk/localizations | 2.0.1-snapshot.v59b1e3b |
| @clerk/nextjs | 5.0.1-snapshot.v59b1e3b |
| @clerk/clerk-react | 5.0.1-snapshot.v59b1e3b |
| @clerk/remix | 4.0.1-snapshot.v59b1e3b |
| @clerk/clerk-sdk-node | 5.0.1-snapshot.v59b1e3b |
| @clerk/shared | 2.0.1-snapshot.v59b1e3b |
| @clerk/themes | 2.0.1-snapshot.v59b1e3b |
| @clerk/types | 4.0.1-snapshot.v59b1e3b |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/backend`
```sh
npm i @clerk/backend@1.0.1-snapshot.v59b1e3b --save-exact
```
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@1.0.1-snapshot.v59b1e3b --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@5.0.1-snapshot.v59b1e3b --save-exact
```
`@clerk/elements`
```sh
npm i @clerk/elements@0.0.2-snapshot.v59b1e3b --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@1.0.1-snapshot.v59b1e3b --save-exact
```
`@clerk/fastify`
```sh
npm i @clerk/fastify@1.0.1-snapshot.v59b1e3b --save-exact
```
`gatsby-plugin-clerk`
```sh
npm i gatsby-plugin-clerk@5.0.1-snapshot.v59b1e3b --save-exact
```
`@clerk/localizations`
```sh
npm i @clerk/localizations@2.0.1-snapshot.v59b1e3b --save-exact
```
`@clerk/nextjs`
```sh
npm i @clerk/nextjs@5.0.1-snapshot.v59b1e3b --save-exact
```
`@clerk/clerk-react`
```sh
npm i @clerk/clerk-react@5.0.1-snapshot.v59b1e3b --save-exact
```
`@clerk/remix`
```sh
npm i @clerk/remix@4.0.1-snapshot.v59b1e3b --save-exact
```
`@clerk/clerk-sdk-node`
```sh
npm i @clerk/clerk-sdk-node@5.0.1-snapshot.v59b1e3b --save-exact
```
`@clerk/shared`
```sh
npm i @clerk/shared@2.0.1-snapshot.v59b1e3b --save-exact
```
`@clerk/themes`
```sh
npm i @clerk/themes@2.0.1-snapshot.v59b1e3b --save-exact
```
`@clerk/types`
```sh
npm i @clerk/types@4.0.1-snapshot.v59b1e3b --save-exact
```

File: packages/elements/package.json [hunk 7]
```diff
@@ -40,7 +49,9 @@
   "main": "./dist/index.js",
   "types": "./dist/index.d.ts",
   "files": [
-    "dist"
+    "dist",
+    "!dist/metafile-*.json",
```
Comment (tmilewski) [inline packages/elements/package.json:53]: Generated for bundle analysis and unable to configure where they're generated. This ensures we don't include them in the resulting package.

File: packages/elements/src/react/common/third-party-providers/social-provider.tsx [hunk 8]
```diff
@@ -33,11 +33,13 @@ export function SocialProvider({ asChild, provider, ...rest }: SocialProviderPro
   }
 
   const Comp = asChild ? Slot : 'button';
+  const defaultProps = asChild ? {} : { type: 'button' as const };
```
Comment (tmilewski) [inline packages/elements/src/react/common/third-party-providers/social-provider.tsx:36]: When encapsulated in form elements, buttons without a type default to `submit` causing issues when hitting enter.

**Review by LekoArts — APPROVED** [review]

**Pull request merged**